File: cocoscreator_assets/assets/scripts/kbe_scripts/Avatar.js
/*-----------------------------------------------------------------------------------------
												entity
-----------------------------------------------------------------------------------------*/

var KBEngine = require("kbengine");

var getCurrentTime = function(){
    　　　　　　 var date = new Date();
                var year=date.getFullYear(); //获取当前年份
                var mon=date.getMonth()+1; //获取当前月份
                var da=date.getDate(); //获取当前日
                var day=date.getDay(); //获取当前星期几
                var hour=date.getHours(); //获取小时
                var minute=date.getMinutes(); //获取分钟
                var second=date.getSeconds(); //获取秒
                var millseocnd =date.getMilliseconds()
                var str = "当前时间: " + hour + ":" + minute + ":" + second + "." +  millseocnd;
                return str;
　　　　};

KBEngine.Avatar = KBEngine.Entity.extend(
    {
        __init__ : function()
        {
            this._super();
            if(this.isPlayer()) {
                KBEngine.Event.fire("enterScene", KBEngine.app.entity_uuid, this.id, this);
            }
        },
                   
        onEnterWorld : function()
        {
            this._super();
            if(this.isPlayer()) {
                KBEngine.INFO_MSG("avatar " +  this.id +" onEnterWorld");
                KBEngine.Event.fire("onAvatarEnterWorld", KBEngine.app.entity_uuid, this.id, this);
            }		
        },

        startWalk: function(moveFlag)
        {
            KBEngine.INFO_MSG("avatar " + this.id + " start walk: " + moveFlag);
            this.cellCall("startWalk", moveFlag);
        },

        onStartWalk: function(moveFlag)
        {
            KBEngine.INFO_MSG("other avatar " + this.id + " on start walk: " + moveFlag);
            KBEngine.Event.fire("otherAvatarOnStartWalk", this.id, moveFlag);
        },

        stopWalk: function(pos)
        {
            cc.log("avatar %d stop walk, pos(%f, %f)", this.id, pos.x, pos.y);   
            var vec3 = new KBEngine.Vector3();
            vec3.x = pos.x;
            vec3.y = pos.y;
            vec3.z = 0.0;
            this.cellCall("stopWalk", pos);
        },

        onStopWalk: function(pos)
        {
            var v2 = new cc.Vec2();
            v2.x = pos.x;
            v2.y = pos.y;
            cc.log("other avatar %d stop walk, pos(%f, %f)", this.id, v2.x, v2.y);   
            KBEngine.Event.fire("otherAvatarOnStopWalk", this.id, v2);
        },

        jump : function()
	    {
            cc.log("avatar %d cell jump", this.id);
		    this.cellCall("jump");
        }, 

        onJump : function()
	    {
            cc.log("other avatar %d on jump", this.id);
		    KBEngine.Event.fire("otherAvatarOnJump", this);
        }, 

        rightJump: function()
        {
            this.cellCall("rightJump");
            KBEngine.INFO_MSG("avatar " + this.id + " right jump");
        },

        onRightJump: function()
        {
            KBEngine.Event.fire("otherAvatarOnRightJump", this.id);
        },

        leftJump: function()
        {
            this.cellCall("leftJump");
            KBEngine.INFO_MSG("avatar " + this.id + " left jump");
        },

        onLeftJump: function()
        {
            KBEngine.Event.fire("otherAvatarOnLeftJump", this.id);
        },

        onPickUpItem : function(itemID, positon)
        {
            var point = new cc.Vec2(positon.x, positon.y);
            KBEngine.Event.fire("otherAvatarOnPickUpItem", this.id, itemID, point);
        },

        pickUpItem : function(itemID, pos)
	    {
            var vec3 = new KBEngine.Vector3();
            vec3.x = pos.x;
            vec3.y = pos.y;
            vec3.z = 0.0;
            this.cellCall("pickUpItem", itemID, vec3);
            console.log("cellCall: avatar pick item = %d pos(%f, %f)", itemID, pos.x, pos.y);
        }, 
                    
        throwItem : function(itemID, force)
        {
            var vec3 = new KBEngine.Vector3();
            vec3.x = force.x;
            vec3.y = force.y;
            vec3.z = 0.0;
            this.cellCall("throwItem", itemID, vec3);
            KBEngine.INFO_MSG("avatar throw time: " + getCurrentTime());
        },
         
        onThrowItem : function(itemID, force)
        {
            var v2 = new cc.Vec2(force.x, force.y);
            KBEngine.Event.fire("otherAvatarThrowItem", this.id, itemID, v2);
            KBEngine.INFO_MSG("another avatar on throw time: " + getCurrentTime());
        },

        onNewTurn : function(eid, second)
	    {
            KBEngine.INFO_MSG("avatar " + eid + " on new turn");
		    KBEngine.Event.fire("newTurn", eid, second);
        }, 

        newTurn : function()
	    {
            this.cellCall("newTurn");
        }, 

        //捡起石头，时间到了，不扔石头，就还原石头的位置
        recoverItem: function(itemID)
        {
            this.cellCall("recoverItem", itemID);
        },

        onRecoverItem: function(itemID)
        {
            KBEngine.Event.fire("otherAvatarRecoverItem", this.id, itemID);
        },

        recvDamage: function(itemID)
        {
            this.cellCall("recvDamage", itemID);
            KBEngine.INFO_MSG("avatar " + this.id + " recvDamage from item=" + itemID);
        },

        onRecvDamage: function(avatarID, harm, hp)
        {
            KBEngine.INFO_MSG("avatar " + avatarID + " recv harm=" + harm + " hp=" + hp);
            KBEngine.Event.fire("onRecvDamage", avatarID, harm, hp);
        },

        onDie: function(avatarID)
        {
            KBEngine.INFO_MSG("avatar " + avatarID + " die");
            KBEngine.Event.fire("onAvatarDie", avatarID);
        },

        onGameOver: function(isWin, hitRate, totalTime, totalHarm, score)
        {
            KBEngine.INFO_MSG("Game is over: avatar " + this.id + "win= " + isWin.toString());
            KBEngine.Event.fire("onGameOver", this.id, isWin, hitRate, totalTime, totalHarm, score);
        },

        //石头出界，重置石头
        resetItem: function(itemID)
        {
            KBEngine.INFO_MSG("reset item ......");
            this.cellCall("resetItem", itemID);
        },

        onResetItem: function(itemID, position)
        {
            KBEngine.INFO_MSG("on reset item position(" + position.x + ", " + position.y + ", " + position.z + ")");
            KBEngine.Event.fire("onResetItem", itemID, position);
        },

        //没石头扔，就产生石头
        addItem: function(left)
        {
            KBEngine.INFO_MSG("add item ......: " + left.toString());
            this.cellCall("addItem", left);
        },

        continueGame: function()
        {
            this.cellCall("continueGame");
            KBEngine.INFO_MSG("avatar " + this.id + " continue game");
        },

        onContinueGame: function(avatarID)
        {
            KBEngine.INFO_MSG("12 avatar " + avatarID + "on continue game, local avatarID=" + this.id);
            KBEngine.Event.fire("onAvatarEnterWorld", KBEngine.app.entity_uuid, this.id, this);

            // if(this.isPlayer()) {
            //     KBEngine.INFO_MSG(" player enter world");
            //     KBEngine.Event.fire("onAvatarEnterWorld", KBEngine.app.entity_uuid, this.id, this);
            // } else {
            //     KBEngine.INFO_MSG(" other enter world");
            //     KBEngine.Event.fire("onEnterWorld", this);
            // }
        }

    });
    
    
    

File: kbengine_stone_assets/scripts/cell/Avatar.py
# -*- coding: utf-8 -*-
import KBEngine
from KBEDebug import *
import GameUtils
import GameConfigs
import random
import copy
from interfaces.EntityCommon import EntityCommon


class Avatar(KBEngine.Entity, EntityCommon):
	def __init__(self):
		KBEngine.Entity.__init__(self)
		EntityCommon.__init__(self)
		DEBUG_MSG("Avatar %i cell new" % (self.id))
		DEBUG_MSG(self.position)
		self.startPosition = copy.deepcopy(self.position)
		self.getCurrRoom().onEnter(self)
		
	def isAvatar(self):
		"""
		virtual method.
		"""
		return True
	#--------------------------------------------------------------------------------------------
	#                              Callbacks
	#--------------------------------------------------------------------------------------------
	def onTimer(self, tid, userArg):
		pass

	def onGetWitness(self):
		"""
		KBEngine method.
		绑定了一个观察者(客户端)
		"""
		DEBUG_MSG("Avatar::onGetWitness: %i." % self.id)

	def onLoseWitness(self):
		"""
		KBEngine method.
		解绑定了一个观察者(客户端)
		"""
		DEBUG_MSG("Avatar::onLoseWitness: %i." % self.id)
	
	def onDestroy(self):
		"""
		KBEngine method.
		entity销毁
		"""
		DEBUG_MSG("Avatar::onDestroy: %i." % self.id)
		room = self.getCurrRoom()
		
		if room:
			room.onLeave(self.id)

	def rightJump(self, exposed):
		DEBUG_MSG("avavtar %i right jump, selfID=%i" % (exposed, self.id))
		if exposed != self.id:
			return
		DEBUG_MSG("avatar %i start right jump" % (self.id))
		self.otherClients.onRightJump()

	def leftJump(self, exposed):
		DEBUG_MSG("avavtar %i left jump, selfID=%i" % (exposed, self.id))
		if exposed != self.id:
			return
		DEBUG_MSG("avatar %i start left jump" % (self.id))
		self.otherClients.onLeftJump()

	def jump(self, exposed):
		"""
		defined.
		玩家跳跃 我们广播这个行为
		"""
		DEBUG_MSG("receive avavtar %i jump, selfID=%i" % (exposed, self.id))
		if exposed != self.id:
			return
		DEBUG_MSG("avatar %i start jump" % (self.id))
		self.otherClients.onJump()

	def pickUpItem(self, exposed, itemID, position):
		"""
		defined.
		玩家捡石头 我们广播这个行为
		"""
		DEBUG_MSG("avavtar %i pickUpItem, selfID=%i" % (exposed, self.id))
		if exposed != self.id:
			return
		DEBUG_MSG("avatar %i pick up item=%i" % (self.id, itemID))
		DEBUG_MSG(position)
		self.otherClients.onPickUpItem(itemID, position)


	def throwItem(self, exposed, itemID, force):
		DEBUG_MSG("avavtar %i throw item=%d, force, selfID=%i" % (exposed, itemID, self.id))
		if exposed != self.id:
			return

		item = None
		room = self.getCurrRoom()
		if room:
			room.killNewTurnTimer()
			item = room.findItemByID(itemID)
			item.throwPlayerID = self.id

		self.throwCount += 1
			
		DEBUG_MSG("avatar %i pick up item=%i" % (self.id, itemID))
		self.otherClients.onThrowItem(itemID, force)

	def newTurn(self, exposed):
		DEBUG_MSG("avavtar %i newTurn, selfID=%i" % (exposed, self.id))
		if exposed != self.id:
			return

		room = self.getCurrRoom()
		
		if room:
			room.nextPlayer()

	def stopWalk(self, exposed, pos):
		DEBUG_MSG("avavtar %i stopWalk, selfID=%i" % (exposed, self.id))
		if exposed != self.id:
			return

		self.otherClients.onStopWalk(pos)


	def startWalk(self, exposed, moveFlag):
		DEBUG_MSG("avavtar %i start walk, selfID=%i moveflag=%i" % (exposed, self.id, moveFlag))
		if exposed != self.id:
			return

		self.otherClients.onStartWalk(moveFlag)

	def recoverItem(self, exposed, itemID):
		DEBUG_MSG("avavtar %i recover item:%i, selfID=%i" % (exposed, itemID, self.id))
		if exposed != self.id:
			return

		self.otherClients.onRecoverItem(itemID)

	def reset(self):
		self.harmCount = 0
		DEBUG_MSG("avatar %i reset: harmCount=%i" % (self.id, self.harmCount))

	def recvDamage(self, exposed, itemID):
		DEBUG_MSG("avavtar %i recvDamage: itemID=%i, selfID=%i, harmCount=%i" % (exposed, itemID, self.id, self.harmCount))
		if exposed != self.id:
			return

		self.harmCount += 1
		if self.harmCount > 1:
			return

		room = self.getCurrRoom()
		otherPlayer = None
		item = None
		harm = 0
		if room:
			item = room.findItemByID(itemID)

		if item:
			harm = item.harm
			if item.throwPlayerID > 0:
				otherPlayer = room.findAvatarByID(item.throwPlayerID)

		if otherPlayer and otherPlayer.id != self.id:
			otherPlayer.hitCount += 1
			otherPlayer.totalHarm += harm

		self.HP = self.HP - harm
		if self.HP <= 0:
			self.HP = 0

		DEBUG_MSG("avatar %i recv harm=%i hp=%i harmCount=%i from player %i " % (exposed, harm, self.HP, self.harmCount, otherPlayer.id))

		self.client.onRecvDamage(self.id, harm, self.HP)
		self.otherClients.onRecvDamage(self.id, harm, self.HP)

		#HP小于等于0，则通知播放死亡动画，然后结束当局游戏
		if self.HP <= 0:
			DEBUG_MSG("avatar id=%i name=%s is defeated !!!" % (self.id, self.name))
			self.client.onDie(self.id)
			self.otherClients.onDie(self.id)
			room.gameOver()


	def isDead(self):
		return self.HP <= 0

	def leaveRoom(self, exposed):
		pass

	def resetItem(self, exposed, itemID):
		DEBUG_MSG("reset item %i" % (itemID))
		room = self.getCurrRoom()
		if room:
			item = room.resetItem(itemID)

	def addItem(self, exposed, left):
		DEBUG_MSG("add item ......")

		room = self.getCurrRoom()
		if room:
			room.addItem(left)

	def resetGameData(self):
		self.hitRate = 0.0
		self.totalTime = 0
		self.totalHarm = 0
		self.score = 0
		self.hitCount = 0
		self.throwCount = 0
		self.HP = 100

	def continueGame(self, exposed):
		DEBUG_MSG("Avatar %i continueGame" % (self.id))
		if self.id != exposed:
			return
		DEBUG_MSG(self.position)
		self.position = copy.deepcopy(self.startPosition)
		DEBUG_MSG(self.startPosition)
		self.client.onContinueGame(self.id)
		#self.otherClients.onContinueGame(self.id)

		room = self.getCurrRoom()
		if room:
			room.addReadyPlayerCount(1, self)
			